### PR TITLE
Update BT nodes to use callback groups / executors to process

### PR DESCRIPTION
**Pull request by BriceRenaudeau**
Instead of spinning the parent node given from the BT blackboard, the BT nodes create callback groups and executor to be pinned. This will allow removing the `_rclcpp_node` in BT manager nodes.

---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #2251  |
| Primary OS tested on | Ubuntu|
| Robotic platform tested on | Wyca simulation and robots |

---

## Description of contribution in a few bullet points

I added 1 callback and 1 executor and replaced the `rclcpp::spin` by `executor.spin` in all BT nodes.

* bt_actino_node
* bt_service_node
* planner_selector
* controller_selector

## Future work that may be required in bullet points

* Once this PR is merged someone (maybe me) will be able to remove `_rclcpp_node` BT parent node from BT manager (#816)

#### For Maintainers: 
- [x] Check that any new parameters added are updated in navigation.ros.org
- [ ] Check that any significant change is added to the migration guide
- [x] Check that any new functions have Doxygen added
- [x] Check that any new features have test coverage
- [x] Check that any new plugins is added to the plugins page
- [x] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

### CODE REVIEW

File: nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp [hunk 1]
```diff
@@ -47,6 +47,10 @@ class BtActionNode : public BT::ActionNodeBase
   : BT::ActionNodeBase(xml_tag_name, conf), action_name_(action_name)
```
Comment (SteveMacenski) [inline nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp:47]: I believe this is missing:
- Goal updater https://github.com/ros-planning/navigation2/blob/main/nav2_behavior_tree/include/nav2_behavior_tree/plugins/decorator/goal_updater_node.hpp#L74
- Battery low https://github.com/ros-planning/navigation2/blob/main/nav2_behavior_tree/include/nav2_behavior_tree/plugins/condition/is_battery_low_condition.hpp#L77
- is stuck https://github.com/ros-planning/navigation2/blob/main/nav2_behavior_tree/include/nav2_behavior_tree/plugins/condition/is_stuck_condition.hpp#L93

File: nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp [hunk 2]
```diff
@@ -47,6 +47,10 @@ class BtActionNode : public BT::ActionNodeBase
...
   {
```
Comment (SteveMacenski) [inline nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp:48]: A note in the migration guide would be great that the BT nodes now have separate executors to ensure deterministic spinning of only the relevant data sources to each BT node

**Review by SteveMacenski — COMMENTED** [review]
No requests in code changes, all good. I think we're just missing a few to call it "done"

File: nav2_behavior_tree/plugins/condition/is_battery_low_condition.cpp [hunk 3]
```diff
@@ -33,10 +33,25 @@ IsBatteryLowCondition::IsBatteryLowCondition(
   getInput("battery_topic", battery_topic_);
   getInput("is_voltage", is_voltage_);
   node_ = config().blackboard->get<rclcpp::Node::SharedPtr>("node");
+  callback_group_ = node_->create_callback_group(
+    rclcpp::CallbackGroupType::MutuallyExclusive,
+    false);
+  callback_group_executor_.add_callback_group(callback_group_, node_->get_node_base_interface());
+  callback_group_executor_thread = std::thread([this]() {callback_group_executor_.spin();});
```
[the remark this answers is not in the document]
Comment (BriceRenaudeau) [reply nav2_behavior_tree/plugins/condition/is_battery_low_condition.cpp:40]: We can get rid of the spinning thread doing a `spin_some` in the `tick()` but we cannot be sure to receive the last battery message. It depends on the battery topic frequency.

[the remark this answers is not in the document]
Comment (SteveMacenski) [reply nav2_behavior_tree/plugins/condition/is_battery_low_condition.cpp]: What do you mean? `spin_some` will spin until the queue is empty. The actual subscriber has a depth, so in worst case situation, we'd only be processing the latest  `N` messages, which are always ordered to my understanding of ROS2. I'm not sure how spinning in a background thread would change that

**Comment by SteveMacenski** [comment]
Can we apply that in the other BT nodes that a thread was added to as well? 

It also looks like you have a merge conflict now. Happy to merge at that point!

**Comment by BriceRenaudeau** [comment]
> Can we apply that in the other BT nodes that a thread was added to as well?

Ok for `goal_updater_node`
This doesn't seem to work for `is_stuck_condition`

**Comment by SteveMacenski** [comment]
Can you clarify on `is_stuck_condition`? What wasn't working? This doesn't seem any different from battery low, takes in a topic, stores some value, recalls it in `tick()`. I don't see anything structurally that would change that. 

Otherwise, this LGTM and is ready to go

**Comment by BriceRenaudeau** [comment]
I don't really understand why it's not working for `is_stuck_condition` but the testing fails.
The odom_callback is not triggered if the callback_group is not spinned in the thread.

**Comment by BriceRenaudeau** [comment]
It's good to go for me but I didn't add the changes to the migration guide.

**Review by SteveMacenski — APPROVED** [review]

**Comment by SteveMacenski** [comment]
@BriceRenaudeau I added https://github.com/ros-planning/navigation.ros.org/blob/master/migration/Foxy.rst

**Comment by SteveMacenski** [comment]
Thanks for your help on this!!

**Pull request merged**